directories.go

```diff
@@ -0,0 +1,134 @@
+package main
+
+import (
+	"errors"
+	"fmt"
+	"github.com/la5nta/pat/internal/debug"
+	"log"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
+
+	"github.com/adrg/xdg"
+)
+
+var lock = &sync.Mutex{}
+var dataPath, configPath, statePath string
+
+func DataDir() string {
+	return getDir(&dataPath, xdg.DataHome, "DataDir")
+}
+
+func ConfigDir() string {
+	return getDir(&configPath, xdg.ConfigHome, "ConfigDir")
+}
+
+func StateDir() string {
+	return getDir(&statePath, xdg.StateHome, "StateDir")
+}
+
+func getDir(dir *string, basePath string, methodName string) string {
+	lock.Lock()
+	defer lock.Unlock()
+	if *dir == "" {
+		initDir(dir, basePath, methodName)
+	}
+	return *dir
+}
+
+func initDir(dir *string, basePath string, methodName string) {
+	*dir = filepath.Join(basePath, strings.ToLower(AppName))
+	if _, err := os.Stat(*dir); os.IsNotExist(err) {
+		err := os.MkdirAll(*dir, os.ModeDir|0755)
+		if err != nil {
+			log.Fatalf("unable to create or open %s %s: %v", methodName, *dir, err)
+		}
+	}
+}
+
+func PrintDirectories() {
+	data := DataDir()
+	cfg := ConfigDir()
+	state := StateDir()
+
+	fmt.Printf("Config directory: \t%s\n", cfg)
+	if data != cfg || state != cfg {
+		fmt.Printf("Mailbox directory:\t%s\n", data)
+		fmt.Printf("Logs directory:   \t%s\n", state)
+	}
+}
+
+func MigrateLegacyDataDir() {
+	homeDir, err := os.UserHomeDir()
+	if err != nil {
+		log.Fatal(err)
+	}
+	legacyDataDir := filepath.Join(homeDir, ".wl2k")
+	f, err := os.Stat(legacyDataDir)
+	switch {
```

Comment:
Please consider using the `switch f, err := os....; {` syntax to make it clear for any reader that this `f` is only handled by the switch cases.

Reply:
Done here and elsewhere to make variable scopes easier to track